### DISCUSSION TITLE
how to combine this two scope into one scope

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1735] [Q&A] [answered]
`the stream scope API is  a little abstract to combine,so how to combine this two scope into one scope`

 private scopeStructMembers(struct: StructsList): Scope {
    const allMembers = this.getStructChain(struct).flatMap((e) => e.items);
    const s = stream(allMembers)
      .map((e) => {
        return this.descriptions.createDescription(e, e.variables);
      })
      .nonNullable();
    return new StreamScope(s);
  }
  
  
  ---------------------------------------------------------------------------------------------------------
  
   private scopeStructMembers(struct: StructsList): Scope {
    const allMembers = this.getStructChain(struct).flatMap((e) => e.items);
    const s = stream(allMembers)
      .map((e) => {
        return this.descriptions.createDescription(e, e.variables);
      })
      .nonNullable();
    const allDecl = struct.items as unknown as Struct_Var_Decl_Init[];
    let hasNextDecl = allDecl.filter((item) => item.nextVariables.length > 0);
    let outerScope;
    hasNextDecl.forEach((item) => {
      let nextVariables = item.nextVariables;
      let nextScope = this.handleNextVariables(item, nextVariables);
      outerScope = new StreamScope(s, nextScope);
    });
    return this.createScopeForNodes(allMembers, outerScope);
  }

**Comment by zrwsmd** [comment]
TYPE
	   Struct111: STRUCT
	      aa,bb,cc,dd: INT;
     END_STRUCT
   END_TYPE

   PROGRAM Test
   VAR
     s1:Struct111
   END_VAR

for my using,the above first struct scope method  can handle s1.aa:=1;(first variables in var declaration)

while above second struct scope method can handle s1.bb:=1; or s1.cc:=1; or s1.dd:=1;
but i want to combine it into one scopeStructMembers method

**Comment by msujew** [comment, marked as answer]
Hey @zrwsmd,

the naive and generic method simply would be to do the following:

```ts
return new StreamScope(scopeA.getAllElements().concat(scopeB.getAllElements()));
```

I would assume there is a solution with better performance, for the specific case.
